File: app/src/main/java/com/example/poste/activities/RegisterActivity.java
package com.example.poste.activities;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;

import com.example.poste.R;

import static com.example.poste.utils.ValidationUtils.validateEmail;
import static com.example.poste.utils.ValidationUtils.validateNames;
import static com.example.poste.utils.ValidationUtils.validatePassword;
import android.content.Intent;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;
import com.example.poste.http.MyApiService;
import com.example.poste.http.RegisterRequest;
import com.example.poste.http.RetrofitClient;
import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class RegisterActivity extends AppCompatActivity {

    private EditText emailEditText;

    private EditText firstNameText;

    private EditText lastNameText;

    private EditText passwordEditText;

    private EditText confirmPasswordEditText;

    private Button registerButton;
    private Button backButton;

    @Override
    public void onBackPressed() {
        Intent back = new Intent(RegisterActivity.this, IntroActivity.class);
        startActivity(back);
        finish();
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_register);

        emailEditText = findViewById(R.id.editTextEmail);
        firstNameText = findViewById(R.id.editTextFirstName);
        lastNameText = findViewById(R.id.editTextLastName);
        passwordEditText = findViewById(R.id.editTextPassword);
        confirmPasswordEditText = findViewById(R.id.editTextConfirmPassword);
        registerButton = findViewById(R.id.buttonRegister);
        backButton = findViewById(R.id.buttonBack);

        registerButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                // Retrieve the values
                String email = emailEditText.getText().toString();
                String firstName = firstNameText.getText().toString().trim();
                String lastName = lastNameText.getText().toString().trim();
                String password = passwordEditText.getText().toString();
                String confirmPassword = confirmPasswordEditText.getText().toString();
                if (!validateEmail(RegisterActivity.this, email)) {
                    return;
                }
                if (!validateNames(RegisterActivity.this, firstName, lastName)) {
                    return;
                }
                if (!validatePassword(RegisterActivity.this, password, confirmPassword)){
                    return;
                }
                MyApiService apiService = RetrofitClient.getRetrofitInstance().create(MyApiService.class);
                Call<ResponseBody> call = apiService.registerUser(new RegisterRequest(email, firstName, lastName, password));
                call.enqueue(new Callback<ResponseBody>() {
                    @Override
                    public void onResponse(Call<ResponseBody> call, Response<ResponseBody> response) {
                        if(response.isSuccessful()) {
                            Toast.makeText(RegisterActivity.this, "Registration successful!", Toast.LENGTH_SHORT).show();
                            // open dashboard activity
                            Intent dashboardIntent = new Intent(RegisterActivity.this, IntroActivity.class);
                            startActivity(dashboardIntent);
                            finish();
                        } else {
                            Toast.makeText(RegisterActivity.this, "Registration failed.", Toast.LENGTH_SHORT).show();
                        }
                    }

                    @Override
                    public void onFailure(Call<ResponseBody> call, Throwable t) {
                        Toast.makeText(RegisterActivity.this, "Registration failed.", Toast.LENGTH_SHORT).show();
                    }
                });
            }
        });
        backButton.setOnClickListener(v -> {
            // Simulate a back press; go back to LandingActivity
            getOnBackPressedDispatcher().onBackPressed();
            finish();
        });
    }
}

File: app/src/main/java/com/example/poste/activities/RegisterActivity.kt
package com.example.poste.activities

import android.os.Bundle
import android.view.View
import android.widget.Button
import com.example.poste.utils.ValidationUtils.validateEmail
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class RegisterActivity : AppCompatActivity() {
    private var emailEditText: EditText? = null
    private var firstNameText: EditText? = null
    private var lastNameText: EditText? = null
    private var passwordEditText: EditText? = null
    private var confirmPasswordEditText: EditText? = null
    private var registerButton: Button? = null
    private var backButton: Button? = null
    protected fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_register)
        getSupportActionBar().hide()
        emailEditText = findViewById(R.id.editTextEmail)
        firstNameText = findViewById(R.id.editTextFirstName)
        lastNameText = findViewById(R.id.editTextLastName)
        passwordEditText = findViewById(R.id.editTextPassword)
        confirmPasswordEditText = findViewById(R.id.editTextConfirmPassword)
        registerButton = findViewById(R.id.buttonRegister)
        backButton = findViewById(R.id.buttonBack)
        registerButton!!.setOnClickListener(View.OnClickListener {
            // Retrieve the values
            val email: String = emailEditText.getText().toString()
            val firstName: String = firstNameText.getText().toString().trim { it <= ' ' }
            val lastName: String = lastNameText.getText().toString().trim { it <= ' ' }
            val password: String = passwordEditText.getText().toString()
            val confirmPassword: String = confirmPasswordEditText.getText().toString()
            if (!validateEmail(this@RegisterActivity, email)) {
                return@OnClickListener
            }
            if (!validateNames(this@RegisterActivity, firstName, lastName)) {
                return@OnClickListener
            }
            if (!validatePassword(this@RegisterActivity, password, confirmPassword)) {
                return@OnClickListener
            }
            val apiService: MyApiService = RetrofitClient.getRetrofitInstance().create(
                MyApiService::class.java
            )
            val call: Call<ResponseBody> =
                apiService.registerUser(RegisterRequest(email, firstName, lastName, password))
            call.enqueue(object : Callback<ResponseBody?> {
                override fun onResponse(
                    call: Call<ResponseBody?>,
                    response: Response<ResponseBody?>
                ) {
                    if (response.isSuccessful()) {
                        Toast.makeText(
                            this@RegisterActivity,
                            getString(R.string.register_success),
                            Toast.LENGTH_SHORT
                        ).show()
                        // open dashboard activity
                        val dashboardIntent =
                            Intent(this@RegisterActivity, IntroActivity::class.java)
                        startActivity(dashboardIntent)
                        finish()
                    } else {
                        Toast.makeText(
                            this@RegisterActivity,
                            getString(R.string.register_failed),
                            Toast.LENGTH_SHORT
                        ).show()
                    }
                }

                override fun onFailure(call: Call<ResponseBody?>, t: Throwable) {
                    Toast.makeText(
                        this@RegisterActivity,
                        getString(R.string.register_failed),
                        Toast.LENGTH_SHORT
                    ).show()
                }
            })
        })
        backButton!!.setOnClickListener { v: View? ->
            // Simulate a back press; go back to LandingActivity
            onBackPressed()
        }
    }
}